### PR TITLE
Feat/add measurement probabilities

### LINKED ISSUE #43: Add a function which, given a ket, returns measurement probabilities

### LINKED ISSUE #59: simulate_shots does not correctly handle complex values
The amplitudes computed in the simulate_shots function are incorrect when kets have nonzero imaginary terms.

For instance, the samples in the following example should be "0" instead of "1":

```
julia> circuit = QuantumCircuit(qubit_count=1, bit_count=0);

julia> push_gate!(circuit, sigma_x(1));

julia> push_gate!(circuit, phase(1));

julia> push_gate!(circuit, sigma_x(1));

julia> push_gate!(circuit, phase(1))
Quantum Circuit Object:
   id: 52530f1c-b6d9-11ed-0fca-b9527064440e 
   qubit_count: 1 
   bit_count: 0 
q[1]:──X────S────X────S──
                         



julia> shots = simulate_shots(circuit, 5)
5-element Vector{String}:
 "1"
 "1"
 "1"
 "1"
 "1"
```

### PR DESCRIPTION
**Pull request by npoirie1**
Adds functions for computing the measurement probabilities of a ket or a circuit. Probabilities for a subset of qubits can also be determined.

Fixes #43 and fixes #59.

### CODE REVIEW

**Comment by npoirie1** [comment]
I found a bug with the get_measurement_probabilities function. When an entry in a ket has real and imaginary parts that are negative, the computed probability is negative. I will fix this issue immediately.

**Comment by npoirie1** [comment]
The bug in the get_measurement_probabilities functions has been fixed. A similar bug in the simulate_shots function has also been fixed.

File: test/qobj.jl [hunk 1]
```diff
@@ -143,4 +143,57 @@ end
...
+end
+
+@testset "change_to_decimal_integer" begin
+    symbols = [0, 1, 2, 1]
+    base = 3
+    decimal = Snowflake.change_to_decimal_integer(symbols, base)
+    @test decimal == 16
+end
+
+@testset "get_ket_index" begin
```
Comment (plerouxanyon) [inline test/qobj.jl:170]: See comment of `change_number_base` test

File: src/core/qobj.jl [hunk 2]
```diff
@@ -661,6 +661,138 @@ function normalize!(x::Ket)
...
+julia> target_qubit = [2];
+
+julia> get_measurement_probabilities(ψ, target_qubit)
+2-element Vector{Float64}:
+ 0.9999999999999998
+ 0.0
+
+```
+"""
+function get_measurement_probabilities(x::Ket)
```
Comment (plerouxanyon) [inline src/core/qobj.jl:710]: Specifying the return type would be useful for a user. It is not ideal if a user should run or look at the code to determine the function's return type.

This applies to all `get_measurement_probabilities` functions.

File: src/core/qobj.jl [hunk 3]
```diff
@@ -661,6 +661,138 @@ function normalize!(x::Ket)
...
+    remaining_bodies::Vector{<:Integer},
+    hspace_size_per_body::Integer, target_index::Integer,
+    remainder_index::Integer)
+
+    target_symbols = change_number_base(target_index, hspace_size_per_body,
+        length(target_bodies))
+    remainder_symbols = change_number_base(remainder_index, hspace_size_per_body,
+        length(remaining_bodies))
+    num_symbols = length(target_bodies)+length(remaining_bodies)
+    combined_symbols = Vector{Int}(undef, num_symbols)
```
Comment (plerouxanyon) [inline src/core/qobj.jl:750]: Using concrete type, `Int`, but dispatching on any `Integer` (`Int8`, `Int16`, ...)

**Review by plerouxanyon — CHANGES_REQUESTED** [review]

**Comment by plerouxanyon** [comment]
Regarding return type annotation in Julia:

@npoirie1  has sent me a link: https://discourse.julialang.org/t/do-you-put-return-type-in-function-definitions/32120/16

There doesn't appear to be a consensus. In this specific case:

* Ket is a generic type and not an abstract type
* The type result is obvious given the type input
* Annotating the return type with `AbstractVector{T}` would not limit the implementation

Considering all this, I think that, in this case, it makes sense to annotate the return type as part of the public interface. However, the return type should still be as abstract as reasonable (`AbtractVector` and not `Vector`, etc.). This won't be limiting, and changing the return type significantly would be a breaking change anyway. Having the return type available is really useful for allowing code to be reasoned about locally without diving into another function's implementation detail.

I think the best would be to handle this on a case-by-case basis.

File: test/qobj.jl [hunk 4]
```diff
@@ -143,4 +143,57 @@ end
 @testset "is_hermitian" begin
     @test is_hermitian(sigma_y())
     @test !is_hermitian(sigma_p())
-end
\ No newline at end of file
+end
+
+@testset "change_number_base" begin
```
[the remark this answers is not in the document]
Comment (npoirie1) [reply test/qobj.jl]: The tests of "private" functions have been removed.

File: test/qobj.jl (hunk 1 above)
Comment (npoirie1) [reply test/qobj.jl]: The tests of "private" functions have been removed.

File: test/qobj.jl [hunk 5]
```diff
@@ -143,4 +143,57 @@ end
...
+    decimal = 88
+    @test_throws ErrorException Snowflake.change_number_base(decimal, base, num_symbols)
+
+    decimal = 16
+    symbols = Snowflake.change_number_base(decimal, base, num_symbols)
+    expected_symbols = [0, 1, 2, 1]
+    @test symbols == expected_symbols
+end
+
+@testset "change_to_decimal_integer" begin
```
[the remark this answers is not in the document]
Comment (npoirie1) [reply test/qobj.jl]: The tests of "private" functions have been removed.

File: src/core/qobj.jl (hunk 3 above)
Comment (npoirie1) [reply src/core/qobj.jl]: Now using the parametric type of the arguments.

**Comment by npoirie1** [comment]
@plerouxanyon, all your comments have been addressed. The get_measurement_probabilities function now accepts a different Hilbert space size for each body.

File: src/core/qobj.jl [hunk 6]
```diff
@@ -682,6 +684,159 @@ function normalize!(x::Ket)
...
+    hspace_size_per_body_list = fill(hspace_size_per_body, num_bodies)
+    return get_measurement_probabilities(x, target_bodies, hspace_size_per_body_list)
+end
+
+function get_measurement_probabilities(x::Ket{Complex{T}},
+    target_bodies::Vector{U},
+    hspace_size_per_body::Vector{U})::AbstractVector{T} where {T<:Real, U<:Integer}
+
+    throw_if_targets_are_invalid(x, target_bodies, hspace_size_per_body)
+    amplitudes = real.(adjoint.(x) .* x)
```
[the remark this answers is not in the document]
Comment (npoirie1) [reply src/core/qobj.jl]: `get_measurement_probabilities` is now reused

**Comment by npoirie1** [comment]
@plerouxanyon, the new comment and the merge conflicts have been addressed.

**Pull request merged**